In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('E:\\Other Stuff\\Python Exercises\\Predicting Pulsars\\pulsar_stars.csv')

In [3]:
data.head(5)

Mean of the integrated profile  \
0                       140.562500   
1                       102.507812   
2                       103.015625   
3                       136.750000   
4                        88.726562   

    Standard deviation of the integrated profile  \
0                                      55.683782   
1                                      58.882430   
2                                      39.341649   
3                                      57.178449   
4                                      40.672225   

    Excess kurtosis of the integrated profile  \
0                                   -0.234571   
1                                    0.465318   
2                                    0.323328   
3                                   -0.068415   
4                                    0.600866   

    Skewness of the integrated profile   Mean of the DM-SNR curve  \
0                            -0.699648                   3.199833   
1                            -0.515088                   1.677258   
2                             1.051164                   3.121237   
3                            -0.636238                   3.642977   
4                             1.123492                   1.178930   

    Standard deviation of the DM-SNR curve  \
0                                19.110426   
1                                14.860146   
2                                21.744669   
3                                20.959280   
4                                11.468720   

    Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
0                              7.975532                      74.242225   
1                             10.576487                     127.393580   
2                              7.735822                      63.171909   
3                              6.896499                      53.593661   
4                             14.269573                     252.567306   

   target_class  
0             0  
1             0  
2             0  
3             0  
4             0

In [4]:
data.target_class.value_counts()

0    16259
1     1639
Name: target_class, dtype: int64

In [5]:
data = data.sort_values('target_class', ascending=False)

In [6]:
data.head(5)

Mean of the integrated profile  \
5207                         37.320312   
11034                        21.828125   
11031                        82.531250   
3851                         55.500000   
3852                         31.796875   

        Standard deviation of the integrated profile  \
5207                                       41.672258   
11034                                      32.080735   
11031                                      42.869704   
3851                                       33.190972   
3852                                       34.805791   

        Excess kurtosis of the integrated profile  \
5207                                     3.754494   
11034                                    5.674437   
11031                                    1.613626   
3851                                     2.799066   
3852                                     4.327562   

        Skewness of the integrated profile   Mean of the DM-SNR curve  \
5207                             14.524742                  81.920569   
11034                            33.193380                 116.206522   
11031                             3.934069                   6.620401   
3851                             13.091231                  30.227425   
3852                             20.432820                  38.005017   

        Standard deviation of the DM-SNR curve  \
5207                                 75.562702   
11034                                83.188276   
11031                                34.381974   
3851                                 70.483271   
3852                                 62.443727   

        Excess kurtosis of the DM-SNR curve   Skewness of the DM-SNR curve  \
5207                               0.738276                      -0.499832   
11034                             -0.001290                      -1.224396   
11031                              5.587993                      30.969379   
3851                               2.201155                       3.303331   
3852                               1.843819                       2.581170   

       target_class  
5207              1  
11034             1  
11031             1  
3851              1  
3852              1

In [7]:
min_max = np.zeros([2, len(data.columns)])
for i in range(len(data.columns)):
    min_max[0, i] = data.iloc[:, i].max()
    min_max[1, i] = data.iloc[:, i].min()

In [8]:
min_max

array([[ 1.92617188e+02,  9.87789107e+01,  8.06952205e+00,
         6.81016217e+01,  2.23392140e+02,  1.10642211e+02,
         3.45398442e+01,  1.19100084e+03,  1.00000000e+00],
       [ 5.81250000e+00,  2.47720418e+01, -1.87601118e+00,
        -1.79188598e+00,  2.13210702e-01,  7.37043217e+00,
        -3.13926961e+00, -1.97697560e+00,  0.00000000e+00]])

In [10]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
just = scaler.fit_transform(np.array(data.iloc[:, 0]).reshape(-1, 1))

In [20]:
np.array(data.iloc[:, 0]).reshape(-1, 1).shape

(17898, 1)

In [14]:
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
just = scaler.fit_transform(np.array(data.iloc[:, 0]).reshape(-1, 1))

In [21]:
new_values = np.zeros([len(data), len(data.columns)])

In [25]:
for i in range(len(data.columns)):
    scaler = MinMaxScaler()
    values = scaler.fit_transform(np.array(data.iloc[:, i]).reshape(-1, 1))
    new_values[:, i] = values.reshape(-1)

C:\Users\user\Anaconda3\lib\site-packages\sklearn\utils\validation.py:475: DataConversionWarning: Data with input dtype int64 was converted to float64 by MinMaxScaler.
  warnings.warn(msg, DataConversionWarning)


In [27]:
df = pd.DataFrame(new_values)

In [29]:
df.columns = data.columns

In [30]:
data.iloc[0, :]

 Mean of the integrated profile                  37.320312
 Standard deviation of the integrated profile    41.672258
 Excess kurtosis of the integrated profile        3.754494
 Skewness of the integrated profile              14.524742
 Mean of the DM-SNR curve                        81.920569
 Standard deviation of the DM-SNR curve          75.562702
 Excess kurtosis of the DM-SNR curve              0.738276
 Skewness of the DM-SNR curve                    -0.499832
target_class                                      1.000000
Name: 5207, dtype: float64

In [32]:
target = df.target_class
values = df.drop(['target_class'], axis=1)

In [33]:
from sklearn.cross_validation import train_test_split

x_train, x_test, y_train, y_test = train_test_split(values, target, test_size=0.2, random_state=50)

C:\Users\user\Anaconda3\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [37]:
from keras.models import Sequential
from keras.layers import Dense, Dropout

In [45]:
model = Sequential()

model.add(Dense(16, activation='relu'))
model.add(Dropout(0.2))

model.add(Dense(4, activation='relu'))
model.add(Dropout(0.2))

model.add(Dense(2, activation='softmax'))

model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['accuracy'])

In [46]:
model.fit(train_x, train_y, epochs=10, batch_size=32)

Instructions for updating:
Use tf.cast instead.
Epoch 1/10
14318/14318 [==============================] - 3s 200us/step - loss: 0.2270 - acc: 0.9273
Epoch 2/10
14318/14318 [==============================] - 1s 77us/step - loss: 0.1354 - acc: 0.9524
Epoch 3/10
14318/14318 [==============================] - 1s 65us/step - loss: 0.1210 - acc: 0.9610
Epoch 4/10
14318/14318 [==============================] - 1s 65us/step - loss: 0.1188 - acc: 0.9599
Epoch 5/10
14318/14318 [==============================] - 1s 58us/step - loss: 0.1156 - acc: 0.9604
Epoch 6/10
14318/14318 [==============================] - 1s 56us/step - loss: 0.1149 - acc: 0.9605
Epoch 7/10
14318/14318 [==============================] - 1s 49us/step - loss: 0.1132 - acc: 0.9610
Epoch 8/10
14318/14318 [==============================] - 1s 44us/step - loss: 0.1120 - acc: 0.9619
Epoch 9/10
14318/14318 [==============================] - 1s 43us/step - loss: 0.1135 - acc: 0.9612
Epoch 10/10
14318/14318 [==========================

In [40]:
train_x = np.array(x_train)
train_y = np.array(y_train)

In [42]:
train_y

array([0., 0., 1., ..., 0., 0., 0.])

In [43]:
from keras.utils import to_categorical

train_y = to_categorical(train_y)

In [48]:
prediction = model.predict_classes(x_test)

In [49]:
prediction

array([0, 0, 0, ..., 0, 1, 0], dtype=int64)

In [51]:
test_y = np.array(y_test)

In [52]:
h = []
for i in range(len(test_y)):
    if prediction[i] == test_y[i]:h.append(1)
    else:h.append(0)

In [55]:
h = np.array(h)
h.sum()/len(h)

0.9784916201117319

After the test, we can see the accuracy of this neural network from the result above. This classifier has an accuracy of 0.97849. 